### PR TITLE
[Arc] Add deduplication pass

### PR DESCRIPTION
**Pull request by fabianschuiki**
Add the `Dedup` pass which deduplicates arcs with identical definitions. If arcs differ only by constant values, the pass outlines the constants such that the arcs can be deduplicated.

Identical arcs are surprisingly common since splitting modules along the fundamental state elements yields a finer-grained subdivision than the modules themselves, which has the potential to uncover additional redundancies.

### CODE REVIEW

File: lib/Dialect/Arc/Transforms/Dedup.cpp [hunk 1]
```diff
@@ -0,0 +1,460 @@
...
+    // Hash the attributes. (Excluded in constant invariant hash.)
+    for (auto namedAttr : op->getAttrDictionary()) {
+      auto name = namedAttr.getName();
+      auto value = namedAttr.getValue();
+      // Skip names.
+      if (name == "sym_name")
+        continue;
+
+      // Hash the interned pointer.
+      unsigned skipConstInvariant = isa<ConstantOp>(op) && name == "value";
```
[the remark this answers is not in the document]
Comment (fabianschuiki) [reply lib/Dialect/Arc/Transforms/Dedup.cpp:107]: Yeah good idea! I think we can just ignore all attributes on const operations for now. That sounds better than relying on specifically the "value" attribute name.

File: lib/Dialect/Arc/Transforms/Dedup.cpp [hunk 2]
```diff
@@ -0,0 +1,460 @@
...
+      auto *opB = values.second.getDefiningOp();
+      if (!opA || !opB) {
+        auto argA = values.first.dyn_cast<BlockArgument>();
+        auto argB = values.second.dyn_cast<BlockArgument>();
+        if (argA && argB) {
+          if (argA.getArgNumber() != argB.getArgNumber())
+            return false;
+        } else {
+          assert(argA || argB);
+          divergences.insert(values);
```
[the remark this answers is not in the document]
Comment (fabianschuiki) [reply lib/Dialect/Arc/Transforms/Dedup.cpp:192]: Excellent catch. We do. Reworked this code a bit to make it borderline more readable.

File: lib/Dialect/Arc/Transforms/Dedup.cpp [hunk 3]
```diff
@@ -0,0 +1,460 @@
...
+      continue;
+    LLVM_DEBUG(llvm::dbgs()
+               << "- Outlining " << outlinedConsts.size() << " consts from "
+               << defineOp.getSymNameAttr() << "\n");
+
+    // Add additional ports to the arc and replace the outlined constants.
+    unsigned newArgOffset = defineOp.getNumArguments();
+    SmallVector<Type> newInputTypes;
+    for (auto type : defineOp.getFunctionType().getInputs())
+      newInputTypes.push_back(type);
```
[the remark this answers is not in the document]
Comment (fabianschuiki) [reply lib/Dialect/Arc/Transforms/Dedup.cpp:387]: Oops :innocent:

File: test/Dialect/Arc/dedup.mlir [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+// CHECK-NOT: arc.define @OutlineConstB
+arc.define @OutlineConstB(%arg0: i4) -> i4 {
+  %c7_i4 = hw.constant 7 : i4
+  %0 = comb.and bin %arg0, %c7_i4 : i4
+  arc.output %0 : i4
+}
+
+// CHECK-LABEL: hw.module @OutlineConst
+hw.module @OutlineConst(%clock: i1, %x: i4, %y: i4) {
```
[the remark this answers is not in the document]
Comment (fabianschuiki) [reply test/Dialect/Arc/dedup.mlir:69]: Excellent point! I added a bunch of new tests for the weird corner cases and rewrote parts of the pass to fix the issues that uncovered :smirk:

File: test/Dialect/Arc/dedup.mlir [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+
+// CHECK-LABEL: hw.module @VariadicDiffsDontDedup
+hw.module @VariadicDiffsDontDedup(%clock: i1, %x: i4, %y: i4, %z: i4) {
+  // CHECK-DAG: arc.state @VariadicDiffsDontDedupA(%x, %y, %z)
+  // CHECK-DAG: arc.state @VariadicDiffsDontDedupB(%x, %y)
+  %0 = arc.state @VariadicDiffsDontDedupA(%x, %y, %z) lat 0 : (i4, i4, i4) -> i4
+  %1 = arc.state @VariadicDiffsDontDedupB(%x, %y) lat 0 : (i4, i4) -> i4
+  // CHECK-NEXT: hw.output
+}
+// CHECK-NEXT: }
```
Comment (maerhart) [inline test/Dialect/Arc/dedup.mlir:228]: These are some great tests!

It probably makes sense to add a few more for these cases:
* one with an `scf.if` or similar in an arc to see regions work properly
* one where (only) different attributes on ops hinder dedup
* one where (only) different types on ops hinder dedup

File: lib/Dialect/Arc/Transforms/Dedup.cpp [hunk 6]
```diff
@@ -0,0 +1,688 @@
+//===- Dedup.cpp ----------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Support/BackedgeBuilder.h"
```
Comment (maerhart) [inline lib/Dialect/Arc/Transforms/Dedup.cpp:10]: Unused?

**Review by maerhart — APPROVED** [review]
Thanks for this big rewrite and all the comments! Cool to see what this pass is able to do.

It seems reasonable to me, but it's not exactly trivial to think about all the corner cases here tbh.

File: test/Dialect/Arc/dedup.mlir (hunk 5 above)
Comment (fabianschuiki) [reply test/Dialect/Arc/dedup.mlir:233]: Excellent suggestions! Added those test. And of course they uncovered an issue where dedup would just blatantly ignore regions... :innocent:

**Comment by maerhart** [comment]
Oops, seems I broke a lot of stuff with adding the `arc.define` region verifier 😬
We definitely want to allow `scf.if` operations inside arcs as long as they contain no operations with side-effects. Not sure what's the best way to do this. Feel free to comment out/revert the region verifier for now to move forward with this.

**Comment by fabianschuiki** [comment]
> Oops, seems I broke a lot of stuff with adding the `arc.define` region verifier

That was a great addition! All I had to do was change it from `isPure` (which `scf.if` doesn't pass because it isn't speculatable -- no idea what that means in the MLIR context) to just `isMemoryEffectFree`. Then `scf.if` even takes care of looking through the ops in its body with the `HasRecursiveMemoryEffects` trait :+1:
